File: services/connect-proxy/flagd_sampler.go
package main

import (
	"context"
	"encoding/binary"
	"log/slog"
	"math"
	"sync/atomic"
	"time"

	"github.com/open-feature/go-sdk/openfeature"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

// FlagdSampler reads trace_sampling_rate from flagd.
// Uses a cached rate updated by a background goroutine.
type FlagdSampler struct {
	cachedRate atomic.Uint64
	client     openfeature.IClient
}

// NewFlagdSampler creates a new sampler that reads from flagd.
func NewFlagdSampler() *FlagdSampler {
	s := &FlagdSampler{
		client: getFlagClient(),
	}
	s.cachedRate.Store(math.Float64bits(1.0)) // default: sample all
	return s
}

// ShouldSample implements sdktrace.Sampler.
func (s *FlagdSampler) ShouldSample(p sdktrace.SamplingParameters) sdktrace.SamplingResult {
	rate := math.Float64frombits(s.cachedRate.Load())

	if rate >= 1.0 {
		return sdktrace.SamplingResult{
			Decision:   sdktrace.RecordAndSample,
			Tracestate: trace.SpanContextFromContext(p.ParentContext).TraceState(),
		}
	}

	if rate <= 0.0 {
		return sdktrace.SamplingResult{
			Decision:   sdktrace.Drop,
			Tracestate: trace.SpanContextFromContext(p.ParentContext).TraceState(),
		}
	}

	// Deterministic sampling based on trace ID
	traceID := p.TraceID
	hash := binary.BigEndian.Uint64(traceID[8:16])
	threshold := uint64(rate * float64(math.MaxUint64))

	decision := sdktrace.Drop
	if hash < threshold {
		decision = sdktrace.RecordAndSample
	}

	return sdktrace.SamplingResult{
		Decision:   decision,
		Tracestate: trace.SpanContextFromContext(p.ParentContext).TraceState(),
	}
}

// Description implements sdktrace.Sampler.
func (s *FlagdSampler) Description() string {
	return "FlagdSampler"
}

// StartRateUpdater spawns a goroutine that polls flagd every 2 seconds.
func (s *FlagdSampler) StartRateUpdater(ctx context.Context) {
	go func() {
		ticker := time.NewTicker(2 * time.Second)
		defer ticker.Stop()

		failing := false
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				rate, err := s.client.FloatValue(ctx, "trace_sampling_rate", 1.0, openfeature.EvaluationContext{})
				if err != nil {
					slog.Debug("flagd: failed to read trace_sampling_rate", "error", err)
					failing = true
					continue
				}
				if failing {
					slog.Info("flagd: trace_sampling_rate recovered", "rate", rate)
					failing = false
				}
				s.cachedRate.Store(math.Float64bits(rate))
			}
		}
	}()
}

File: services/rust-hid/src/flagd_sampler.rs
//! FlagdSampler — Dynamic trace sampling via OpenFeature/flagd.
//!
//! Reads the `trace_sampling_rate` flag (float 0.0–1.0) from flagd
//! on every sampling decision. Falls back to 1.0 (sample all) if
//! flagd is unavailable.

use opentelemetry::trace::{Link, SpanKind, TraceId, TraceState};
use opentelemetry_sdk::trace::{ShouldSample, SamplingDecision, SamplingResult};
use opentelemetry::KeyValue;
use std::sync::atomic::{AtomicU64, Ordering};

/// Shared cached sampling rate between the reader and background updater.
static CACHED_RATE: AtomicU64 = AtomicU64::new(u64::MAX); // sentinel for "not set"

/// Sampler that reads trace_sampling_rate from flagd.
#[derive(Debug, Clone, Default)]
pub struct FlagdSampler;

impl FlagdSampler {
    pub fn new() -> Self {
        Self
    }
}

impl ShouldSample for FlagdSampler {
    fn should_sample(
        &self,
        _parent_context: Option<&opentelemetry::Context>,
        trace_id: TraceId,
        _name: &str,
        _span_kind: &SpanKind,
        _attributes: &[KeyValue],
        _links: &[Link],
    ) -> SamplingResult {
        let rate = get_sampling_rate_sync();

        if rate >= 1.0 {
            return SamplingResult {
                decision: SamplingDecision::RecordAndSample,
                attributes: Vec::new(),
                trace_state: TraceState::default(),
            };
        }

        if rate <= 0.0 {
            return SamplingResult {
                decision: SamplingDecision::Drop,
                attributes: Vec::new(),
                trace_state: TraceState::default(),
            };
        }

        // Deterministic sampling based on trace ID for consistency
        let trace_id_bytes = trace_id.to_bytes();
        let hash = u64::from_be_bytes(trace_id_bytes[8..16].try_into().unwrap_or([0; 8]));
        let threshold = (rate * u64::MAX as f64) as u64;

        let decision = if hash < threshold {
            SamplingDecision::RecordAndSample
        } else {
            SamplingDecision::Drop
        };

        SamplingResult {
            decision,
            attributes: Vec::new(),
            trace_state: TraceState::default(),
        }
    }
}

/// Synchronous wrapper to get sampling rate from flagd.
/// Uses tokio::Runtime::block_on since ShouldSample is sync.
fn get_sampling_rate_sync() -> f64 {
    let cached = CACHED_RATE.load(Ordering::Relaxed);
    if cached != u64::MAX {
        return f64::from_bits(cached);
    }

    // Default to 1.0 (sample all) — the async updater will set the real value
    1.0
}

/// Spawn a background task that periodically updates the cached sampling rate.
/// Call this once during initialization.
pub fn spawn_rate_updater() {
    tokio::spawn(async move {
        loop {
            let rate = get_sampling_rate_async().await;
            CACHED_RATE.store(rate.to_bits(), Ordering::Relaxed);
            tokio::time::sleep(std::time::Duration::from_secs(2)).await;
        }
    });
}

async fn get_sampling_rate_async() -> f64 {
    use open_feature::{EvaluationContext, OpenFeature};

    let of = OpenFeature::singleton().await;
    let client = of.create_client();
    let ctx = EvaluationContext::default();
    client
        .get_float_value("trace_sampling_rate", Some(&ctx), None)
        .await
        .unwrap_or(1.0)
}
